File: core/claude.py
"""
Async Claude API wrapper.

Uses streaming + get_final_message() to handle large outputs without timeout.
Adaptive thinking is enabled for Sonnet/Opus only (not Haiku).
"""

from __future__ import annotations
import json
import re
import anthropic
from config import THINKING_MODELS

_client = anthropic.AsyncAnthropic()


async def call_claude(
    model: str,
    system: str,
    messages: list[dict],
    max_tokens: int = 16_000,
) -> str:
    """Call Claude and return the text response."""
    kwargs: dict = {
        "model": model,
        "max_tokens": max_tokens,
        "system": system,
        "messages": messages,
    }
    if model in THINKING_MODELS:
        kwargs["thinking"] = {"type": "adaptive"}

    async with _client.messages.stream(**kwargs) as stream:
        response = await stream.get_final_message()

    for block in response.content:
        if block.type == "text":
            return block.text

    raise ValueError(f"Claude ({model}) returned no text content")


def parse_json(raw: str) -> dict:
    """
    Extract and parse JSON from a Claude response.
    Handles markdown code fences and leading/trailing prose.
    """
    cleaned = raw.strip()

    # Strip markdown code fences
    cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"\s*```$", "", cleaned)

    # Advance to first { or [
    first_brace   = cleaned.find("{")
    first_bracket = cleaned.find("[")

    if first_brace == -1 and first_bracket == -1:
        raise ValueError("No JSON object found in Claude response")

    if first_brace != -1 and (first_bracket == -1 or first_brace < first_bracket):
        start = first_brace
    else:
        start = first_bracket

    cleaned = cleaned[start:]

    # Find matching closing delimiter
    open_c  = cleaned[0]
    close_c = "}" if open_c == "{" else "]"
    depth, end = 0, -1
    in_string = False
    escape_next = False

    for i, ch in enumerate(cleaned):
        if escape_next:
            escape_next = False
            continue
        if ch == "\\" and in_string:
            escape_next = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == open_c:
            depth += 1
        elif ch == close_c:
            depth -= 1
            if depth == 0:
                end = i
                break

    if end != -1:
        cleaned = cleaned[: end + 1]

    return json.loads(cleaned)

File: src/utils/claude.ts
import Anthropic from '@anthropic-ai/sdk';

const client = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

export const MODEL = 'claude-opus-4-6';

/**
 * Call Claude with a system prompt and messages, returning the full text response.
 * Uses streaming internally to handle large outputs without timeout.
 */
export async function callClaude(
  systemPrompt: string,
  messages: Array<{ role: 'user' | 'assistant'; content: string }>,
  maxTokens = 16000
): Promise<string> {
  const stream = client.messages.stream({
    model: MODEL,
    max_tokens: maxTokens,
    thinking: { type: 'adaptive' },
    system: systemPrompt,
    messages,
  });

  const response = await stream.finalMessage();

  // Extract text from response content blocks
  const textBlocks = response.content.filter((b) => b.type === 'text');
  if (textBlocks.length === 0) {
    throw new Error('Claude returned no text content');
  }
  return textBlocks.map((b) => (b as { type: 'text'; text: string }).text).join('');
}

/**
 * Parse JSON from a Claude response string.
 * Claude sometimes wraps JSON in markdown code blocks — this strips them.
 */
export function parseJSON<T>(raw: string): T {
  // Strip markdown code blocks: ```json ... ``` or ``` ... ```
  let cleaned = raw.trim();
  cleaned = cleaned.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/i, '');

  // Extract the outermost JSON object or array
  const firstBrace = cleaned.indexOf('{');
  const firstBracket = cleaned.indexOf('[');
  let start = -1;

  if (firstBrace !== -1 && (firstBracket === -1 || firstBrace < firstBracket)) {
    start = firstBrace;
  } else if (firstBracket !== -1) {
    start = firstBracket;
  }

  if (start > 0) {
    cleaned = cleaned.slice(start);
  }

  // Find matching closing brace/bracket
  const openChar = cleaned[0];
  const closeChar = openChar === '{' ? '}' : ']';
  let depth = 0;
  let end = -1;
  for (let i = 0; i < cleaned.length; i++) {
    if (cleaned[i] === openChar) depth++;
    else if (cleaned[i] === closeChar) {
      depth--;
      if (depth === 0) {
        end = i;
        break;
      }
    }
  }

  if (end !== -1) {
    cleaned = cleaned.slice(0, end + 1);
  }

  return JSON.parse(cleaned) as T;
}
